Library/Homebrew/cmd/list.rb

```diff
@@ -153,21 +169,41 @@ def filtered_list(args:)
       pinned_versions = {}
       names.sort.each do |d|
         keg_pin = (HOMEBREW_PINNED_KEGS/d.basename.to_s)
-        pinned_versions[d] = keg_pin.readlink.basename.to_s if keg_pin.exist? || keg_pin.symlink?
+        next if !keg_pin.exist? && !keg_pin.symlink?
+
+        keg = keg_pin.realpath
+        next unless select_keg?(keg, args)
+
+        pinned_versions[d] = keg.basename.to_s
       end
       pinned_versions.each do |d, version|
         puts d.basename.to_s.concat(args.versions? ? " #{version}" : "")
       end
     else # --versions without --pinned
       names.sort.each do |d|
-        versions = d.subdirs.map { |pn| pn.basename.to_s }
+        versions = d.subdirs.map { |pn| pn.basename.to_s if select_keg?(pn, args) }.compact
+        next if versions.empty?
         next if args.multiple? && versions.length < 2
 
         puts "#{d.basename} #{versions * " "}"
       end
     end
   end
 
+  def select_formula?(formula, args)
+    select_tab?(Tab.for_formula(formula), args)
+  end
+
+  def select_keg?(keg, args)
+    select_tab?(Tab.for_keg(keg), args)
+  end
+
+  def select_tab?(tab, args)
+    return true if !args.requested? && !args.unrequested?
+
+    [!!args.requested?, !args.unrequested?].include?(!!tab.installed_on_request)
```

Comment:
Having to use `!!` is a bit weird and confusing here. Could you explain what's going on and/or rewrite this to make it easier to understand? Using `#present?` or `#blank?` might help.

Reply:
Also, ideally any `args.x?` method would always return a `Boolean`. That could be handled in another PR, though.

Reply:
They should already do so after #10880, unless the option starts with `--[no-]` in which case it's intentionally triple state.